add random tube phantom (with -N <num>)

jtamir opened the pull request:
Adds the ability to make a tube phantom with random number of tubes. Also supports basis.

Example:  
`bart phantom -N 100  -r 123  -x 256 phn`
produces
![image](https://user-images.githubusercontent.com/8716651/89721305-dd53dc00-d990-11ea-834c-8cdbdfb3f1bf.png)

uecker commented on src/simu/phantom.c:
```diff
@@ -517,18 +518,98 @@ static void separate_bckgrd(int Nb, struct ellipsis_s bckgrd[Nb], int Nf, struct
+
+				bool overlap = true;
+
+				while (overlap) {
```
Doesn't this go into an endless loop when N is too high and there is not enough space?

uecker replied:
I would just error out after a reasonable number of trials.

jtamir commented:
It can happen. I have a version that shrinks the tubes over time until they fit. Should I add a graceful fail at n<N tubes if runtime exceeds some number?

Jon
www.jtsense.com
________________________________
From: Martin Uecker <notifications@github.com>
Sent: Sunday, August 9, 2020 9:12:00 AM
To: mrirecon/bart <bart@noreply.github.com>
Cc: Jon Tamir <jtamir@utexas.edu>; Author <author@noreply.github.com>
Subject: Re: [mrirecon/bart] add random tube phantom (with -N <num>) (#231)


@uecker commented on this pull request.

________________________________

In src/simu/phantom.c<https://github.com/mrirecon/bart/pull/231#discussion_r467588506>:

>  {
        if (1 < dims[COEFF_DIM]) {

+               struct ellipsis_bs phantom_tubes_N[2 * N - 1];
+
+               if (random) {
+
+                       // generate random ellipse
+                       for (int i = 0; i < 2 * N - 2; i+=2) {
+
+                               double sx = 0;
+                               double sy = 0;
+                               double px = 0;
+                               double py = 0;
+
+                               bool overlap = true;
+
+                               while (overlap) {


Doesn't this go into an endless loop when N is too high and there is not enough space?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/mrirecon/bart/pull/231#pullrequestreview-463877365>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACCQC23UCRNXM3UXWAZRCA3R72VDBANCNFSM4PY37N7A>.

This message is from an external sender. Learn more about why this matters.<https://ut.service-now.com/sp?id=kb_article&number=KB0011401>

jtamir commented:
I added some checks and changed the way ellipses are added over time. It will now error if it isn't able to find a solution. Nonetheless, with the change I was able to generate 10,000 ellipses (see screenshot):
![image](https://user-images.githubusercontent.com/8716651/89745795-f4ff9300-da6a-11ea-9bf9-37075b48addb.png)

uecker commented:
I will merge this, it is a nice feature! But as these are all spheres at the moment and not ellipses, the check for overlap should be simple, or I am missing something?

jtamir commented:
I wanted to leave it in so that at some point it could be extended, but I think simpler is better. So I refactored, fixed some bug with the borders, and made it explicitly for circles only.

Here's 10,000 tubes, with fixes
![image](https://user-images.githubusercontent.com/8716651/89808966-cb3c7f80-daef-11ea-9990-f021acd6239e.png)
